src/rvsdg/schema.egg

```diff
@@ -48,6 +49,97 @@
 (function Gamma (Operand VecOperand VecVecOperand) Body) ;; branching
 (function Theta (Operand VecOperand VecOperand) Body) ;; loop
 
+;; Substitution
+
+;;                   e  [ x -> v ]
+(function SubstExpr (Expr i64 Operand) Expr)
+(function SubstOperand (Operand i64 Operand) Operand)
+(function SubstBody (Body i64 Operand) Body)
+(function SubstVecOperand (VecOperand i64 Operand) VecOperand)
+(function SubstVecVecOperand (VecVecOperand i64 Operand) VecVecOperand)
+
+(rewrite (SubstExpr (Const ty ops lit) x v) (Const ty ops lit))
+(rewrite (SubstExpr (Call ty f args) x v) (Call ty f (SubstVecOperand args x v)))
+(rewrite (SubstExpr (add ty a b) x v) (add ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (sub ty a b) x v) (sub ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (mul ty a b) x v) (mul ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (div ty a b) x v) (div ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (eq ty a b) x v) (eq ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (lt ty a b) x v) (lt ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (gt ty a b) x v) (gt ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (le ty a b) x v) (le ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (ge ty a b) x v) (ge ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (not ty a b) x v) (not ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (and ty a b) x v) (and ty (SubstOperand a x v) (SubstOperand b x v)))
+(rewrite (SubstExpr (or ty a b) x v) (or ty (SubstOperand a x v) (SubstOperand b x v)))
+
+(rewrite (SubstOperand (Arg x) x v) v)
+(rule ((= f (SubstOperand (Arg y) x v)) (!= y x))
+      ((union f (Arg y))))
+(rewrite (SubstOperand (Node b) x v) (Node (SubstBody b x v)))
+(rewrite (SubstOperand (Project i b) x v) (Project i (SubstBody b x v)))
+
+(rewrite (SubstBody (PureOp e) x v) (PureOp (SubstExpr e x v)))
+;; Subst doesn't cross regions - so we subst into the inputs but not outputs
+;; Node that a Gamma node's idx is on the outside, so it gets affected, but not
+;; a Theta node's predicate
+(rewrite (SubstBody (Gamma pred inputs outputs) x v) (Gamma (SubstOperand pred x v) (SubstVecOperand inputs x v) outputs))
+(rewrite (SubstBody (Theta pred inputs outputs) x v) (Theta pred (SubstVecOperand inputs x v) outputs))
+
+;; params: vec, var, op, next index to subst
+;; rtjoa: TODO: implement by mapping internally so they're not O(n^2) time
```

Comment:
Why is it n^2 time?

Reply:
Because `vec-set` is linear, and we set indices one-at-a-time (this is what I mentioned in person, about how we should eventually either use a vec & built-in map, or a cons-list & map rules, but not mix)

Reply:
Okay I understand thanks